src/GEError.ts

```diff
@@ -0,0 +1,118 @@
+export enum GEErrorEnviornmentSource {
+  CSM = 'CSM',
+  PAM = 'PAM',
+  FRONTEND = 'FRONTEND',
+  RADIO = 'RADIO',
+  OTHER = 'OTHER'
+}
+
+export type GEErrorJSON = {
+  status: number
+  message: string
+  name: string
+  source: GEErrorEnviornmentSource
+  stack?: string
+}
+export class GEError extends Error {
+  readonly status: number
+  readonly name: string
+  readonly source: GEErrorEnviornmentSource
+  readonly stack?: string
+
+  constructor(
+    message: string,
+    status: number,
+    source: GEErrorEnviornmentSource,
+    name: string,
+    stack?: string
+  ) {
+    super(message)
+
+    this.name = name
+    this.status = status
+    this.source = source
+    if (stack) {
+      this.stack = stack
+    }
+    Object.setPrototypeOf(this, new.target.prototype)
+  }
+
+  public toJSON(): GEErrorJSON {
+    return {
+      status: this.status,
+      message: this.message,
+      name: this.name,
+      source: this.source,
+      stack: this.stack
+    }
+  }
+
+  static fromJSON(rawData: GEErrorJSON) {
+    return new GEError(rawData.message, rawData.status, rawData.source, rawData.name, rawData.stack)
+  }
+}
+/*****************************************
+ * PAM
+ ****************************************/
+
+// these error codes draw inspiration from HTTP error codes
+// https://httpstatuses.com/
+
+export enum GEPAMErrorCodes {
+  UNKOWN_CLIENT_TYPE = 415, // ex: known client types are JSONRPC, and serial expected
+  ADD_CLIENT_ERROR = 412, // adding a pam client fails
+  SET_ERROR = 510, // a set action fails
+  RADIO_UNINITIALIZED = 407, // cannot perform an action on the radio because it hasnt been initialized
+  KILL_CLIENT_SESSION_ERROR = 417, // cannot kill radio client session
+  LOGIN_FAILED = 401, // login attempt did not work
+  LOGIN_SUCCEEDED_BUT_NOT_NEEDED = 302, // login requested to radio, but wasnt needed to begin with
+  UNSUPPORTED_ACTION_TYPE = 501, // ex: GET, INIT, and SET are valid action types
+  INVALID_ACTION = 400, // a supported action type did not have the information it was supposed to
+  NETWORK_ERROR = 500, // could not reach the radio
+  NOT_LOGGED_IN = 403 // you are not currently authenticated with the radio
+}
+export class GEPAMError extends GEError {
+  /* istanbul ignore next */
+  constructor(message: string, status: GEPAMErrorCodes) {
+    super(message, status, GEErrorEnviornmentSource.PAM, 'GEPAMError')
+
+    /* istanbul ignore next */
+    Object.setPrototypeOf(this, new.target.prototype)
+  }
+
+  toJSON() {
+    return super.toJSON()
+  }
+}
+
+/*****************************************
+ * CSM
+ ****************************************/
+// these error codes draw inspiration from HTTP error codes
+// https://httpstatuses.com/
+export enum GECSMErrorCodes {
+  NO_FORWARDING_ADDRESS = 421
```

Comment:
The one possibility that I can think of where we might run into an error in the CSM is if the ActionObject string is invalid. This is a slightly different error because we cannot return the error in the action object, we're just going to have to reject with an error and break. Should we just use the built in ActionObject error string?